File: exercises/build_tree_from_traversal_array_105.ts
// Optimal Solution using stop param
function buildTree(preorder: number[], inorder: number[]): TreeNode | null {
	let i = 0;
    let p = 0;
	function solve(preorder: number[], inorder: number[], stop: number): TreeNode | null {
		if (p >= preorder.length) return null;
		if (inorder[i] === stop) { i += 1; return null; }
		const node = new TreeNode(preorder[p]);
		p += 1;
		node.left = solve(preorder, inorder, node.val);
		node.right = solve(preorder, inorder, stop);
		return node;
	}

	return solve(preorder, inorder, -Infinity);
};

// Original Solution not optimal using hashmap
//function buildTree(preorder: number[], inorder: number[]): TreeNode | null {
//    let pre = 0;
//	const map = new Map();
//	function solve(preorder: number[], inorder: number[], start: number, end: number) {
//		if (start == end) return null;
//		const ind = map.get(preorder[pre]);
//		pre += 1;
//		const node = new TreeNode(inorder[ind]);
//		node.left = solve(preorder, inorder, start, ind);
//		node.right = solve(preorder, inorder, ind + 1, end);
//		return node;
//	}
//	// Build map and call helper function
//	for (let i = 0; i < inorder.length; i++) {
//		map.set(inorder[i], i);
//	}
//	return solve(preorder, inorder, 0, preorder.length);
//};

File: exercises/build_tree_from_traversal_array_105.java
// Optimal Solution and using a stop value
class Solution {
    private int in = 0;
    private int pre = 0;
    
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return build(preorder, inorder, Integer.MIN_VALUE);
    }
    
    private TreeNode build(int[] preorder, int[] inorder, int stop) {
        if (pre >= preorder.length) return null;
        if (inorder[in] == stop) {
            in++;
            return null;
        }
        TreeNode node = new TreeNode(preorder[pre]);
        pre += 1;
        node.left = build(preorder, inorder, node.val);
        node.right = build(preorder, inorder, stop);
        return node;        
    }
}
// Inefficient Solution using array slices (not worth it)
// But easier to remember maybe
//class Solution {
//	int pre = -1;
//    public TreeNode buildTree(int[] preorder, int[] inorder) {
//        if (inorder.length == 0) return null;
//        this.pre += 1;
//        int ind = indexOf(inorder, preorder[pre]);
//        TreeNode node = new TreeNode(inorder[ind]);
//        node.left = buildTree(preorder, Arrays.copyOfRange(inorder, 0, ind));
//        node.right = buildTree(preorder, Arrays.copyOfRange(inorder, ind+1, inorder.length));
//        return node;
//    }
//    int indexOf(int[] array, int number){
//        int index = 0;
//        for(int n: array){
//            if(number == n) return index;
//            else index++;
//        }
//        return -1;
//    }
//}

